libbpfgo.h

```diff
@@ -16,35 +16,36 @@ int libbpf_print_fn(enum libbpf_print_level level, const char *format,
   if (level != LIBBPF_WARN)
     return 0;
 
-  // NOTE: va_list args is managed in libbpf caller
-  // however, these copies must be matched with va_end() in this function
-  va_list exclusivity_check, cgroup_check;
-  va_copy(exclusivity_check, args);
-  va_copy(cgroup_check, args);
+  char str[300];
```

Comment:
[geyslan](https://github.com/geyslan) [4 hours ago](https://github.com/aquasecurity/libbpfgo/pull/270#discussion_r1044595970)
@haoxiangbd First, thanks for bringing this up.

I would like to propose an approach in which we don't need a buffer.

We would have to be more precise and check the format first, so if it's a match we could access the correct arguments without fear of accessing the wrong types. For instance, take a look at this kernel selftest code: https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/prog_tests/bpf_tcp_ca.c#L218-L238

[haoxiangbd] I think that's doable but a bit hacky. In the selftest you mentioned, it's assuming that any pr_warn with a format message "-- BEGIN PROG LOAD LOG --" will also have two arguments in the list and both are of type string format. I looked up libbpf code and the two such instances both follow this scheme. However, it someone add a third instance of pr_warn in the future with "-- BEGIN PROG LOAD LOG --" format and a non-string argument, this selftest will crash.
If we really want to go with this path (assuming no future conflicted change), I can code it that way. But I just wanted to double check to see what's wrong with using a temporary buffer to hold the formatted string? libbpf log is not a frequent path, the added string copy is an overhead but it's definitely not that bad.

For the "Esclusivity flag on" we should base our format check on this call: https://github.com/libbpf/libbpf/blob/68e6f83f223ebf3fbf0d94c0f4592e5e6773f0c1/src/nlattr.c#L192
[haoxiangbd] This looks good to me. If we want to go this path, this would work for the time being.

For the "cgroup" and "Invalid argument" we need to find out which format is being passed.
[haoxiangbd] I looked up libbpf code but it's not obvious where those are logged. Can someone who knows this better provide input?

It would also be amazing to have a selftest for this callback.
[haoxiangbd] I can look into this.

WDYT?

Reply:
> I think that's doable but a bit hacky. In the selftest you mentioned, it's assuming that any pr_warn with a format message "-- BEGIN PROG LOAD LOG --" will also have two arguments in the list and both are of type string format. I looked up libbpf code and the two such instances both follow this scheme. However, it someone add a third instance of pr_warn in the future with "-- BEGIN PROG LOAD LOG --" format and a non-string argument, this selftest will crash.

We can rely on **strcmp()**: `if (!strcmp(format, "Kernel error message: %s\n")) {`, so we can detect it on our selftest if the format changes. By using the static format we are making it clear how many args we can access.

> If we really want to go with this path (assuming no future conflicted change), I can code it that way. But I just wanted to double check to see what's wrong with using a temporary buffer to hold the formatted string? libbpf log is not a frequent path, the added string copy is an overhead but it's definitely not that bad.

I assume by not using the buffer it might make the purpose of the callback clearer, since the format will be explicated in the code.

> This looks good to me. If we want to go this path, this would work for the time being.

I would greatly appreciate it.
 
> > For the "cgroup" and "Invalid argument" we need to find out which format is being passed.
> 
> I looked up libbpf code but it's not obvious where those are logged. Can someone who knows this better provide input?

You can temporarily change it to print the format and trigger it using this selftest: https://github.com/aquasecurity/libbpfgo/blob/580127258d2377e1b32622b587707c6f518ac859/selftest/cgroup-legacy/main.go

> > It would also be amazing to have a selftest for this callback.
> 
> I can look into this.

Thank you again for your helping on this.

Reply:
@haoxiangbd if you don't have time for tackling it, I can take it, right? Let me know.

Reply:
> @haoxiangbd if you don't have time for tackling it, I can take it, right? Let me know.

Yes, please help looking into this. I don't have any context on how and when the "cgroup" message is raised.